### DISCUSSION TITLE
The Insatiable Postgres Replication Slot - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#303] [Announcements]
# The Insatiable Postgres Replication Slot - Gunnar Morling

While working on a demo for processing change events from Postgres with Apache Flink,
I noticed an interesting phenomenon:
A Postgres database which I had set up for that demo on Amazon RDS, ran out of disk space.
The machine had a disk size of 200 GiB which was fully used up in the course of less than two weeks.

Now a common cause for this kind of issue are replication slots which are not advanced:
in that case, Postgres will hold on to all WAL segments after the latest log sequence number (LSN) which was confirmed for that slot.
Indeed I had set up a replication slot (via the Decodable CDC source connector for Postgres, which is based on Debezium).
I then had stopped that connector, causing the slot to become inactive.
The problem was though that I was really sure that there was no traffic in that database whatsoever!
What could cause a WAL growth of ~18 GB/day then?

https://www.morling.dev/blog/insatiable-postgres-replication-slot/

**Comment by newbie01github** [comment]
Is there nothing/something that you can setup on AWS that check/drop inactive replication log if it has reach a specific size of some sort?

**Reply by gunnarmorling** [reply]
Not sure whether it's exposed on RDS, but Postgres itself has a setting [max_slot_wal_keep_size](https://postgresqlco.nf/doc/en/param/max_slot_wal_keep_size/) since version 13. But keep in mind that a consumer may not be able to resume to read from the slot after a downtime if any required WAL segment has been removed in the meantime due to surpassing that value.

**Reply by ankur-kr-singh** [reply]
@gunnarmorling Does this mean that once max_slot_wal_keep_size retained wal size is reached then consumer won't read anything from replication slot as it will become inactive ? Suppose parameter is set to 1024 MB, it reaches that but now suppose actual events come in database which needs to be consumed then what will happen ? will it be added in WAL by slot ? I think replication slot would become inactive and consumer won't read anything

**Reply by ankur-kr-singh** [reply]
My main doubt in above is whether after reaching max_slot_wal_keep_size ,  old wal retained records are flushed out automatically so that new wal can be retained ?

**Reply by gunnarmorling** [reply]
When the configured value is reached, older WAL segments, which were prevented from being deleted for usage by the slot, may be deleted by the database. The slot will then become invalid. I.e. the database won't keep up using more and more diskspace to keep WAL for inactive replication slots.

**Comment by cjbooms** [comment]
I am happy to report that this issue has now been patched in the PostgreSQL JDBC driver.
https://engineering.zalando.com/posts/2023/11/patching-pgjdbc.html

**Reply by gunnarmorling** [reply]
Thanks a lot, this is really great!
